src/common/Microsoft.DotNet.UpgradeAssistant.Abstractions/ProjectExtensions.cs

```diff
@@ -58,6 +66,145 @@ public static async ValueTask<bool> AppliesToProjectAsync(this Type type, IProje
+
+            foreach (var applicableLanguage in supportedLanguages)
+            {
+                if (applicableLanguage is not null && !string.IsNullOrWhiteSpace(applicableLanguage))
+                {
+                    if (project.Language == Language.CSharp
+                        && applicableLanguage.Equals(CodeAnalysis.LanguageNames.CSharp, StringComparison.Ordinal))
+                    {
+                        return false;
+                    }
+
+                    if (project.Language == Language.VisualBasic
+                        && applicableLanguage.Equals(CodeAnalysis.LanguageNames.VisualBasic, StringComparison.Ordinal))
+                    {
+                        return false;
+                    }
+
+                    if (project.Language == Language.FSharp
+                        && applicableLanguage.Equals(CodeAnalysis.LanguageNames.FSharp, StringComparison.Ordinal))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Checks to see if the type has an ApplicableLanguageAttribute that should filter the feature out of the user's view.
+        /// </summary>
+        /// <param name="type">a type that represents an upgrade-assistant feature decorated with <see cref="ApplicableLanguageAttribute"/>.</param>
+        /// <param name="project">the project currently being upgraded.</param>
+        /// <returns>true if the object should be hidden from the user. Will return false for objects not defning the language they support.</returns>
+        private static bool DoesApplicableLanguageFilterThisFeature(Type type, IProject project)
+        {
+            var applicableLangAttr = type.CustomAttributes.FirstOrDefault(a => a.AttributeType.FullName.Equals(typeof(ApplicableLanguageAttribute).FullName, StringComparison.Ordinal));
+            if (applicableLangAttr is null)
+            {
+                // this object is not the type we thought it was, the filter does not apply
+                return false;
+            }
+
+            var languageNumbers = applicableLangAttr.ConstructorArguments.First().Value as IEnumerable<System.Reflection.CustomAttributeTypedArgument>;
+            if (languageNumbers is null)
+            {
+                // this object is not the type we thought it was, the filter does not apply
+                return false;
+            }
+
+            foreach (var applicableLangInt in languageNumbers.Select(l => l.Value as int?))
+            {
+                if (!applicableLangInt.HasValue)
+                {
+                    // this object is not the type we thought it was, the filter does not apply
+                    return false;
+                }
+
+                var applicableLanguage = (Language)applicableLangInt;
+                if (project.Language == Language.CSharp && applicableLanguage == Language.CSharp)
```

Comment:
Maybe just check `project.Language == applicableLanguage` instead of having three separate conditions?